File: src/main/java/de/ayont/lpc/channels/ChannelManager.java
package de.ayont.lpc.channels;

import de.ayont.lpc.LPC;
import de.ayont.lpc.storage.Storage;
import de.ayont.lpc.storage.impl.InMemoryStorage;
import de.ayont.lpc.storage.impl.MySQLStorage;
import de.ayont.lpc.storage.impl.SQLiteStorage;
import net.kyori.adventure.text.Component;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;
import org.bukkit.Bukkit;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.logging.Level;

public class ChannelManager {

    private final LPC plugin;
    private final Map<String, Channel> channels = new HashMap<>();
    private final Map<UUID, String> playerChannels = new HashMap<>();
    private String defaultChannel;

    public ChannelManager(LPC plugin) {
        this.plugin = plugin;
    }

    public void init() {
        loadChannels();
    }
    
    public void shutdown() {
    }

    private void initStorage() {
        // Storage is now managed by LPC.java
    }

    public void loadChannels() {
        channels.clear();
        ConfigurationSection section = plugin.getConfig().getConfigurationSection("channels.list");
        if (section == null) return;

        defaultChannel = plugin.getConfig().getString("channels.default-channel", "global");

        for (String key : section.getKeys(false)) {
            ConfigurationSection channelSection = section.getConfigurationSection(key);
            if (channelSection == null) continue;

            String type = channelSection.getString("type", "global").toLowerCase();
            Channel channel;

            switch (type) {
                case "permission":
                    channel = new PermissionChannel(plugin, key, channelSection);
                    break;
                case "range":
                    channel = new RangeChannel(plugin, key, channelSection);
                    break;
                case "global":
                default:
                    channel = new GlobalChannel(plugin, key, channelSection);
                    break;
            }
            channels.put(key, channel);
            plugin.getLogger().info("Loaded channel: " + key);
        }
    }

    public Channel getChannel(String id) {
        return channels.get(id);
    }
    
    public void loadPlayerChannel(Player player) {
        if (plugin.getStorage() == null) return;
        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
            String channelId = plugin.getStorage().load(player.getUniqueId());
            if (channelId != null && channels.containsKey(channelId)) {
                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    playerChannels.put(player.getUniqueId(), channelId);
                });
            }
        });
    }

    public Channel getPlayerChannel(Player player) {
        String channelId = playerChannels.getOrDefault(player.getUniqueId(), defaultChannel);
        return channels.getOrDefault(channelId, channels.get(defaultChannel));
    }

    public void setPlayerChannel(Player player, String channelId) {
        if (channels.containsKey(channelId)) {
            playerChannels.put(player.getUniqueId(), channelId);
            if (plugin.getStorage() != null) {
                plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
                   plugin.getStorage().save(player.getUniqueId(), channelId); 
                });
            }
        }
    }
    
    public Map<String, Channel> getChannels() {
        return channels;
    }

    public void sendMessage(Player sender, String message) {
        // Check for symbol prefix
        for (Channel channel : channels.values()) {
            if (channel.getSymbol() != null && !channel.getSymbol().isEmpty() && message.startsWith(channel.getSymbol())) {
                if (channel.canJoin(sender)) {
                    String finalMessage = message.substring(channel.getSymbol().length()).trim();
                    if (!finalMessage.isEmpty()) {
                        channel.sendMessage(sender, finalMessage);
                        return;
                    }
                }
            }
        }

        // Send to current channel
        Channel current = getPlayerChannel(sender);
        if (current != null) {
            current.sendMessage(sender, message);
        } else {
            sender.sendMessage(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(plugin.getConfig().getString("channels.messages.no-channel", "<red>No channel available.")));
        }
    }
}

File: src/main/kotlin/de/ayont/lpc/channels/ChannelManager.kt
package de.ayont.lpc.channels

import de.ayont.lpc.LPC
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

class ChannelManager(private val plugin: LPC) {

    private val channels = mutableMapOf<String, Channel>()
    private val playerChannels = mutableMapOf<UUID, String>()
    private var defaultChannel: String = "global"

    fun init() {
        loadChannels()
    }
    
    fun shutdown() {
    }

    fun loadChannels() {
        channels.clear()
        val section = plugin.config.getConfigurationSection("channels.list") ?: return

        defaultChannel = plugin.config.getString("channels.default-channel", "global")!!

        for (key in section.getKeys(false)) {
            val channelSection = section.getConfigurationSection(key) ?: continue

            val type = channelSection.getString("type", "global")!!.lowercase()
            val channel: Channel = when (type) {
                "permission" -> PermissionChannel(plugin, key, channelSection)
                "range" -> RangeChannel(plugin, key, channelSection)
                else -> GlobalChannel(plugin, key, channelSection)
            }
            channels[key] = channel
            plugin.logger.info("Loaded channel: $key")
        }
    }

    fun getChannel(id: String): Channel? {
        return channels[id]
    }
    
    fun loadPlayerChannel(player: Player) {
        val storage = plugin.lpcStorage ?: return
        Bukkit.getScheduler().runTaskAsynchronously(plugin, Runnable {
            val channelId = storage.load(player.uniqueId)
            if (channelId != null && channels.containsKey(channelId)) {
                Bukkit.getScheduler().runTask(plugin, Runnable {
                    playerChannels[player.uniqueId] = channelId
                })
            }
        })
    }

    fun getPlayerChannel(player: Player): Channel? {
        val channelId = playerChannels.getOrDefault(player.uniqueId, defaultChannel)
        return channels[channelId] ?: channels[defaultChannel]
    }

    fun setPlayerChannel(player: Player, channelId: String) {
        if (channels.containsKey(channelId)) {
            playerChannels[player.uniqueId] = channelId
            val storage = plugin.lpcStorage
            if (storage != null) {
                Bukkit.getScheduler().runTaskAsynchronously(plugin, Runnable {
                   storage.save(player.uniqueId, channelId) 
                })
            }
        }
    }
    
    fun getChannels(): Map<String, Channel> {
        return channels
    }

    fun sendMessage(sender: Player, message: String) {
        // Check for symbol prefix
        for (channel in channels.values) {
            val symbol = channel.getSymbol()
            if (symbol != null && symbol.isNotEmpty() && message.startsWith(symbol)) {
                if (channel.canJoin(sender)) {
                    val finalMessage = message.substring(symbol.length).trim()
                    if (finalMessage.isNotEmpty()) {
                        channel.sendMessage(sender, finalMessage)
                        return
                    }
                }
            }
        }

        // Send to current channel
        val current = getPlayerChannel(sender)
        if (current != null) {
            current.sendMessage(sender, message)
        } else {
            sender.sendMessage(MiniMessage.miniMessage().deserialize(plugin.config.getString("channels.messages.no-channel", "<red>No channel available.")!!))
        }
    }
}
